internal/impl/gcp/output_pubsub_test.go

```diff
@@ -0,0 +1,244 @@
+
+	client.On("Topic", "test_foo").Return(fooTopic).Once()
+	client.On("Topic", "test_bar").Return(barTopic).Once()
+
+	out, err := newPubSubOutput(conf)
+	require.NoError(t, err, "failed to create output")
+	out.client = client
+	t.Cleanup(func() {
+		err = out.Close(ctx)
+		require.NoError(t, err, "closing output failed")
+
+		mock.AssertExpectationsForObjects(t, client, fooTopic, barTopic)
+	})
+
+	err = out.WriteBatch(ctx, service.MessageBatch{service.NewMessage([]byte("foo"))})
+	require.ErrorContains(t, err, "topic 'test_foo' does not exist")
+
+	err = out.WriteBatch(ctx, service.MessageBatch{service.NewMessage([]byte("bar"))})
+	require.ErrorContains(t, err, "failed to validate topic 'test_bar': simulated error")
+}
+
+func TestPubSubOutput_PublishErrors(t *testing.T) {
+	ctx := context.Background()
+
+	conf, err := newPubSubOutputConfig().ParseYAML(`
+    project: sample-project
+    topic: test_${! content().string().split("_").index(0) }
+    `,
+		nil,
+	)
+	require.NoError(t, err, "bad output config")
+
+	client := &mockPubSubClient{}
+
+	fooTopic := &mockTopic{}
+	fooTopic.On("Exists").Return(true, nil).Once()
+	fooTopic.On("Stop").Return().Once()
+
+	barTopic := &mockTopic{}
+	barTopic.On("Exists").Return(true, nil).Once()
+	barTopic.On("Stop").Return().Once()
+
+	client.On("Topic", "test_foo").Return(fooTopic).Once()
+	client.On("Topic", "test_bar").Return(barTopic).Once()
+
+	fooMsgA := service.NewMessage([]byte("foo_a"))
+	fooResA := &mockPublishResult{}
+	fooResA.On("Get").Return("", errors.New("simulated foo error")).Once()
+	fooTopic.On("Publish", "foo_a", mock.Anything).Return(fooResA).Once()
+
+	fooMsgB := service.NewMessage([]byte("foo_b"))
+	fooResB := &mockPublishResult{}
+	fooResB.On("Get").Return("foo_b", nil).Once()
+	fooTopic.On("Publish", "foo_b", mock.Anything).Return(fooResB).Once()
+
+	barMsg := service.NewMessage([]byte("bar"))
+	barRes := &mockPublishResult{}
+	barRes.On("Get").Return("", errors.New("simulated bar error")).Once()
+	barTopic.On("Publish", "bar", mock.Anything).Return(barRes).Once()
+
+	out, err := newPubSubOutput(conf)
+	require.NoError(t, err, "failed to create output")
+	out.client = client
+	t.Cleanup(func() {
+		err = out.Close(ctx)
+		require.NoError(t, err, "closing output failed")
+
+		mock.AssertExpectationsForObjects(
+			t,
+			client,
+			fooTopic, barTopic,
+			fooResA, fooResB, barRes,
+		)
+	})
+
+	err = out.Connect(ctx)
+	require.NoError(t, err, "connect failed")
+
+	err = out.WriteBatch(ctx, service.MessageBatch{fooMsgA, fooMsgB, barMsg})
+	require.Error(t, err, "did not get expected publish error")
+
+	var batchErr *service.BatchError
+	require.ErrorAs(t, err, &batchErr, "error is not a batch error")
+	require.Equal(t, 2, batchErr.IndexedErrors(), "did not receive expected number of batch errors")
+
+	var errs []string
+	batchErr.WalkMessages(func(i int, m *service.Message, err error) bool {
+		errs = append(errs, err.Error())
+		return true
+	})
```

Comment:
I can't quite figure out why WalkMessages isn't iterating over the two failed messages despite `batchErr.IndexedErrors()` returning 2...

Reply:
Solved in #1837

Reply:
Removed my cherry-picked commit since the fix is in main now. Thanks a lot @Jeffail!